src/main/java/dev/openfeature/sdk/HookContext.java

```diff
@@ -4,6 +4,7 @@
 import lombok.NonNull;
 import lombok.Value;
 import lombok.With;
+import lombok.experimental.NonFinal;
 
 /**
  * A data class to hold immutable context that {@link Hook} instances use.
```

Comment:
Isn't immutable anymore. Which does have some implications. I've not looked at the implementation yet, but a hook implementation could reasonable expect to hold a reference to this context and now it could mutate.

Given this class my initial expectation would be for this to be cloned for each hook invocation with an updated context.

 `hookContext.withNewEvaluationConetxt(EvaluationContext context)` which would return a new instance with a new immutable field.

Like dotnet: https://github.com/open-feature/dotnet-sdk/blob/bb4f3526c2c2c2ca48ae61e883d6962847ebc5a6/src/OpenFeature/Model/HookContext.cs#L69C33-L69C57

Reply:
You're right.

The challenge is at least in a couple hooks, we use the HookContext itself as a WeakMap key. If the HookContext changes across the evaluation, this functionality doesn't work. WeakMaps in hooks are especially useful for maintaining state across stages (before/after).

Reply:
We don't explicitly make any guarantees around either of these in the spec.... but the weakmap use-case in hooks is so useful it would be good to have some consistent reference across stages.

Reply:
So, we should update these comments and make it explicit that the content can really only be trusted for the duration of your invocation. 

Though I do think the weak map situation represents a specification weakness that should be addressed with a specific way to have invocation data. As is it depends on implementation details, and conversely places unexpected limitations on the implementation.

Reply:
Agree that we should update comments at a minimum. I feel a bit stuck, because I don't like this mutability, but on the other hand see a lot of value in the consistent reference.

Do you think it would be silly to have some kind of explicit reference handle? We could instantiate some dedicated object for this in the HookContext and never change it if we wanted, and even make it an optional point in the spec.

Reply:
I think an explicit handle of some kind is a lot better than pulling double duty with another field. It makes it something explicit instead of an implementation detail and makes it obvious to maintainers. The idea of a weak map itself is somewhat technology specific though, so I am not sure it is great for the spec. I am not sure how I would do something equivalent in functional/immutable languages. 

I think, ideally, we would add a specific way to store and retrieve data in the hook invocation. 

I tried it in the LaunchDarkly implementation hooks: https://github.com/launchdarkly/open-sdk-specs/tree/main/specs/HOOK-hooks

Example usage in C#: https://github.com/launchdarkly/dotnet-core/blob/10045f0f97c0f923d23fa6a7c3af0bd701d2ba42/pkgs/telemetry/src/TracingHook.cs#L155

https://github.com/launchdarkly/dotnet-core/blob/10045f0f97c0f923d23fa6a7c3af0bd701d2ba42/pkgs/telemetry/src/TracingHook.cs#L168

It does have a bit of overhead though as you have to at least create 1 object per hook per evaluation. Though it isn't much overhead compared to the other transformations that can happen, like the context ones.

Reply:
I really like the `SeriesData` concept.

I don't know of any Java Hooks currently using WeakMaps (just JS ones) so I'm inclined to go with your initial suggestion and remove the mutability I introduced, and also implement some kind of storage in another PR. If it's sensible we can think about bringing it into the spec.

:pray:

Reply:
I've removed this mutability, I create a new hook-context now from the result of the "before".